sdk/core/System.ClientModel/src/Message/BinaryContent.cs

```diff
@@ -191,4 +203,79 @@ public override void Dispose()
             }
         }
     }
+
+    private sealed class StreamBinaryContent : BinaryContent
+    {
+        private const int CopyToBufferSize = 81920;
+
+        private readonly Stream _stream;
+        private readonly long _origin;
+
+        public StreamBinaryContent(Stream stream)
+        {
+            if (!stream.CanSeek)
+            {
+                throw new ArgumentException("Stream must be seekable.", nameof(stream));
+            }
+
+            _stream = stream;
+            _origin = stream.Position;
+        }
+
+        public override bool TryComputeLength(out long length)
+        {
+            if (_stream.CanSeek)
+            {
+                length = _stream.Length - _origin;
+                return true;
+            }
+
+            length = 0;
+            return false;
+        }
+
+        public override void WriteTo(Stream stream, CancellationToken cancellationToken)
+        {
+            _stream.Seek(_origin, SeekOrigin.Begin);
+
+            // This is not using CopyTo so that we can honor cancellations.
```

Comment:
Can we put this helper into a separate source file/directory with things we wish BCL had and possibly things that we will expose polyfill APIs for.

Reply:
Yep - I think we have the folder - I'll look at moving this other there